src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs

```diff
@@ -563,6 +565,18 @@ private async Task<bool> EnsureConnectionStateAsync(HttpConnectionContext connec
             // Setup the connection state from the http context
             connection.User = connection.HttpContext?.User;
 
+            // TODO: option to enable/disable this
+            var authorizeAttribute = connection.HttpContext?.GetEndpoint()?.Metadata.GetMetadata<AuthorizeAttribute>();
+
+            // Don't check token if the endpoint doesn't have authorization applied to it
+            if (authorizeAttribute is not null)
+            {
+                // AuthenticateResult should be cached since the context has already been authenticated.
+                // This is just to get the AuthenticateResult
+                var authResult = await connection.HttpContext!.AuthenticateAsync(authorizeAttribute.AuthenticationSchemes);
```

Comment:
This doesn't work if you're using policies instead. You'd have to fetch the policy service. @HaoK 

It's probably time to make the auth ticket available as a feature like we do for the user. Knowing how to re-invoke auth to get the ticket again is a pain.

Reply:
https://github.com/dotnet/aspnetcore/issues/20847

cc @HaoK @blowdart

Reply:
Right, well in the meantime, you just need to compute the policy from the authorizeAttribute like here, and call Authenticate on the policy evaluator https://github.com/dotnet/aspnetcore/blob/52eff90fbcfca39b7eb58baad597df6a99a542b0/src/Security/Authorization/Policy/src/AuthorizationMiddleware.cs#L60

Reply:
We don't need to hack it here if we can add the new API as part of this PR. Does it seem reasonable though?

Reply:
So I'm not exactly sure I fully understand the proposal in 20847 and how that would address this problem, since 20847 looks like its talking about adding a feature to store the additional metadata that is equivalent to the User.  But the setting of user is only for the 'default authentication scheme', Its not really directly related to Authorize, as Authorize attributes don't affect the user.  So for example, you can set the default scheme to be Cookies, so User = cookie, and authorize JWT, which wouldn't affect the user at all, it would be really weird if you have User = cookie, with the other properties from whatever the last call to the Authorize service was.  I'm not sure I'm explaining the problem well, but does that make sense?

Reply:
It makes sense to store the result.Properties of the default authenticate scheme along side the context.User, but that is not what I think the code here wants, it wants the authResult properties for an arbitrary Authorize attribute, this is not only talking about the default scheme.

Reply:
It wants the entire `AuthenticateResult`

Reply:
Right, but is it enough to only have the AuthenticateResult for context.User?  That won't help for an Authorize policy that specifies its own authentication scheme and doesn't rely on the default authentication scheme

Reply:
This conversation doesn't really make sense to me (but I'm no auth expert), you're saying the `User` for a request wont be associated with the "Jwt" auth that an endpoint requires if "Cookie" auth is the default scheme for the app?

Reply:
Yes that's the subtle point I was trying to make.  `AddAuthentication("ThisSchemeUserIsOnContext")` vs [Authorize("ThisDoesNotAffectContextUser")].  Only the 'default' sceheme sets the user, as that behavior comes from the AuthenticationMiddleware.  AuthorizationMiddleware does not muck with anything, it only deals with authentication/authorization in the context of handling an AuthZ call.

Reply:
What about `[Authorize(AuthenticationSchemes = "something")]`, that should affect the User right? I understand policy shouldn't, and as far as I know we aren't doing anything with the policy in the current code.

Reply:
No I don't think so, the root AuthZ constructs are just schemes/requirements, so just specifying a scheme means use that scheme and the 'default' authZ policy's requirements.  Unless SignalR is doing something differently, its the AuthenticationMiddleware responsibility is to set the ambient User on the HttpContext.  I think a good way to look at it is, `[Authorize(no scheme specified)]` means do AuthZ using the ambient context.User (but it doesn't control how that user was set, that comes from the defaultScheme being set).  When authorize specifies a scheme or policy, context.User is not used, instead a the ClaimsPrincipal for this AuthZ operation is generated (union of AuthenticateAsync from all requested schemes)

Reply:
Another way to put it, setting context.User or the defaultScheme, has no effect on any `[Authorize(scheme | policy = "foo"]` endpoints

Reply:
but context.User gets updated here so they should be in sync:
https://github.com/dotnet/aspnetcore/blob/73d9fa93e109474982db9fdd52d309c9313296de/src/Security/Authorization/Policy/src/PolicyEvaluator.cs#L52

Reply:
Duh, ok I'm getting old and senile, right I moved the code that used to do all this from AuthorizeFilter into PolicyEvaluator so they all do the same thing...  Okay, never mind, yeah the proposal should be fine then, if we add the new feature that lets you set everything at once, things will be in sync.